modules/packages/EpochManager.chpl

```diff
@@ -0,0 +1,1147 @@
+      var del_epoch = tok.local_epoch.read();
+      if (del_epoch == 0) {
+        writeln("Bad local epoch! Please pin! Using global epoch!");
+        del_epoch = global_epoch.read();
+      }
+      limbo_list[del_epoch].push(x);
+    }
+
+    /*
+      Try to announce a new epoch. If successful, reclaim objects which are
+      safe to reclaim
+    */
+    proc try_reclaim() {
+      var count = EBR_EPOCHS;
+
+      // if nothing to reclaim, try the next epoch, but loop only for one
+      // full cycle
+      while (count) {
+        count = count - 1;
+
+        // Set a flag to let other tasks know that a task is already
+        // trying to reclaim
+        if (is_setting_epoch.testAndSet()) {
+          return;
+        }
+        var reclaim_epoch = try_advance();
+        if (reclaim_epoch == 0) {
+          // try_advance failed
+          is_setting_epoch.clear();
+          return;
+        }
+
+        var reclaim_limbo_list = limbo_list[reclaim_epoch];
+        var head = reclaim_limbo_list.pop();
+        is_setting_epoch.clear();
+
+        while (head != nil) {
+          var next = head.next;
+          delete head.val;
+          reclaim_limbo_list.retire_node(head);
+          head = next;
+        }
+      }
+    }
+
+    /*
+      Reclaim all objects
+    */
+    proc deinit() {
+      delete allocated_list;
+      delete free_list;
+      delete limbo_list;
+    }
+  }
+
+  pragma "no doc"
+  class _token {
+    var local_epoch : atomic uint;
+    var is_registered : atomic bool;
+
+    proc init() {
+    }
+  }
+
+  /*
+    Handle to :class:`EpochManager`
+  */
+  class TokenWrapper {
+
+    pragma "no doc"
+    var _tok : unmanaged _token;
+
+    pragma "no doc"
+    var manager : unmanaged EpochManager;
+
+    pragma "no doc"
+    proc init(_tok : unmanaged _token, manager : unmanaged EpochManager) {
+      this._tok = _tok;
+      this.manager = manager;
+    }
+
+    /*
+      `Pin` a task
+    */
+    proc pin() {
+      manager.pin(this._tok);
+    }
+
+    /*
+      `Unpin` a task
+    */
+    proc unpin() {
+      manager.unpin(this._tok);
+    }
+
+    /*
+      Delete an object.
+
+      :arg x: The class instance to be deleted. Must be of unmanaged class type
+    */
+    proc delete_obj(x) {
+      manager.delete_obj(this._tok, x);
+    }
+
+    /*
+      Try to announce a new epoch. If successful, reclaim objects which are
+      safe to reclaim
+    */
+    proc try_reclaim() {
+      manager.try_reclaim();
+    }
+
+    /*
+      Unregister the handle from the manager
+    */
+    proc unregister() {
+      manager.unregister(this._tok);
+      this._tok = nil;
+    }
+
+    /*
+      Unregister the handle from the manager
+    */
+    proc deinit() {
+      this.unregister();
+    }
+  }
+
+  use VectorModule;
+
+  /*
+    :record:`DistributedEpochManager` manages reclamation of objects, ensuring
+    thread-safety. It employs privatization.
```

Comment:
I'm not so certain that in the eventual more ideal implementation that there would be any difference between DistributedEpochManager and a local one. Couldn't the user use one API for both and the implementation figure out if it is distributed? I understand that might not be practical right now, and this is more a philisophical question for future direction rather than for this PR.

But, if you agree that is the desired future direction, a `.. note::` about it would be appropriate here.

Reply:
> I'm not so certain that in the eventual more ideal implementation that there would be any difference between DistributedEpochManager and a local one.

The `EpochManager` is optimized for shared-memory, as the `DistributedEpochManager` is privatized and performs a lot of additional logic for each compute node. I'd believe that a separate `DistributedEpochManager` is needed. Perhaps a more explicit `SharedMemoryEpochManager` name is needed to differentiate the two, but the single-locale optimizations are important performance-wise.

Reply:
Right, if using the single-node one is an optimization, then shouldn't the "default" be distributed? I.e. EpochManager is distributed and we have also a LocalEpochManager or something?

Reply:
Yeah, that's not a bad idea. @dgarvit what do you think?

Reply:
If you agree @dgarvit we should rename `DistributedEpochManager` to `EpochManager` and the old `EpochManager` to `LocalEpochManager`. I'm wondering if we should utilize some `local` blocks inside of it as well for performance reasons.